### ISSUE TITLE
How to generate voxels for a batch of points?

### ISSUE THREAD

**Issue by Cenbylin** [traveller59/spconv#188]
Thank you all authors.

It seems that `spconv.utils.VoxelGenerator` can only generate voxels from the same group of points.

Do you have any suggestions for speeding up `VoxelGenerator` for a batch of points (with shape [B, N, 3])?

**Comment by Divadi** [comment]
Another question - is it possible to generate voxels w/ torch tensors, preserving gradient flow?

**Comment by Cenbylin** [comment]
Hi @Divadi 
I guess the use of NumPy (instead of Torch) is because of the convenience of c++ binding or operations on Cuda. 

There is still a way to preserve gradient flow. It looks like to re-connect the new voxel tensor and the original compute graph.

You can try this:
```python
import spconv
import torch
# ------------------ #
#  Data Preparation  #
# ------------------ #
ori_points = torch.randn(100, 3, requires_grad=True)
voxel_generator = spconv.utils.VoxelGeneratorV2(voxel_size=[0.1] * 3,
                                                point_cloud_range=[-1, -1, -1, 1, 1, 1],
                                                max_num_points=50,
                                                max_voxels=8000,
                                                full_mean=False)
# ------------------------------- #
#  Associated with compute graph  #
# ------------------------------- #
# 1. retain raw indice. 
ori_points_ = ori_points.detach()
batch_point_indice = torch.arange(0, ori_points_.shape[0]).float().unsqueeze(1)  # [N, 1]
# [N,3] -> [N,4], the add '1' is the indice in N
ori_points_ = torch.cat((ori_points_, batch_point_indice), dim=1).view(-1, 4)

# 2. generate voxels
# features[num_voxels, point_per_voxel, 3+1] 1(indice), coords: [num_voxels, 1+3] 1(batch)
res= voxel_generator.generate(ori_points_.numpy())
# features: [num_voxels, max_num_points_per_voxel, 3]
gen_fea = torch.tensor(res['voxels'], dtype=torch.float32)
# coors: [num_voxels, 1+3]
gen_coords = torch.tensor(res['coordinates'], dtype=torch.int32)
batch_idx = torch.zeros((gen_coords.shape[0], 1), dtype=torch.int32)  # batch=0
gen_coords = torch.cat((batch_idx, gen_coords), dim=1)
# num_points: [num_voxels]
gen_npoints = torch.tensor(res['num_points_per_voxel'], dtype=torch.int32)


# 3. reconnect with compute graph
gen_fea[:,:,:3] = ori_points[gen_fea[:,:,3].long(),:]
# features[num_voxels, point_per_voxel, 3], N.B. set pos value to 0 in indice 0 (filling value)
gen_fea = gen_fea[:,:,:3]


# 4. you can do anything
# voxel_vfe: [num_voxels, 3]
voxel_vfe = torch.mean(gen_fea, dim=1)
voxel_feature = spconv.SparseConvTensor(features=voxel_vfe, indices=gen_coords, 
                                        spatial_shape=[20, 20, 20], 
                                        batch_size=1)
simple_net = spconv.SparseConv3d(in_channels=3, out_channels=6,
                                 kernel_size=3, stride=2)
output = simple_net(voxel_feature).dense()

# ------ #
#  Test  #
# ------ #
loss = output.sum()
print('previous:', ori_points.grad)
loss.backward()
print('current:', ori_points.grad)
```
Hope it can help you.

There is another concern: the points in voxels didn't necessarily cover all original points. So the supervising signal is also sparse (i.e. many zeros in ori_points.grad).

Are you training a point cloud GAN? If so, could we discuss them a little bit?

**Comment by traveller59** [comment]
generate batch of voxels requires cuda hashtable because dense table takes too much GPU memory, I plan to work on it after sparse convolution speed up, may be one month later.
@Divadi Do you mean use VoxelGenerator as a differentiable layer? if so, I can finish it in this weekend.

**Comment by Cenbylin** [comment]
@traveller59 thanks :)

**Comment by Divadi** [comment]
@Cenbylin Really nice solution, thank you!

@traveller59 Yup, but it looks like the solution above is very fast (little to no overhead time even on a large point cloud), so it's not high-priority anymore

**Comment by FrancescoMandru** [comment]
What is spatial shape @Cenbylin ?? I'm struggling to find the right one, I'm adapting my code to your

```
    ptc = load_velo_scan(data[0])[:, 0:3]
    lab = load_velo_label(data[0])
    #
    ptc, lab = random_sampling_with_labs(ptc, lab, 0.1)

    # ------------------ #
    #  Data Preparation  #
    # ------------------ #
    #ori_points = torch.randn(100, 3, requires_grad=True)
    xyzmin = np.floor(np.percentile(ptc, 0, axis=0)).astype(np.int32)
    xyzmax = np.ceil(np.percentile(ptc, 100, axis=0)).astype(np.int32)
    ori_points = torch.tensor(ptc, requires_grad=True)
    voxel_generator = spconv.utils.VoxelGeneratorV2(voxel_size=[0.5] * 3,
                                                    point_cloud_range=[xyzmin[0], xyzmin[1], xyzmin[2], xyzmax[0], xyzmax[1], xyzmax[2]],
                                                    max_num_points=50,
                                                    max_voxels=20000,
                                                    full_mean=False)

    # ------------------------------- #
    #  Associated with compute graph  #
    # ------------------------------- #
    # 1. retain raw indice.
    ori_points_ = ori_points.detach()
    batch_point_indice = torch.arange(0, ori_points_.shape[0]).float().unsqueeze(1)  # [N, 1]
    # [N,3] -> [N,4], the add '1' is the indice in N
    ori_points_ = torch.cat((ori_points_, batch_point_indice), dim=1).view(-1, 4)
    # 2. generate voxels
    # features[num_voxels, point_per_voxel, 3+1] 1(indice), coords: [num_voxels, 1+3] 1(batch)
    # res contains:
    #   voxels: [M, max_points, ndim] float tensor. only contain points.
    #   coordinates: [M, 3] int32 tensor. zyx format.
    #   num_points_per_voxel: [M] int32 tensor.
    res = voxel_generator.generate(ori_points_.numpy())

    # https://github.com/traveller59/spconv/issues/142
    # https://github.com/traveller59/spconv/issues/188

    # features: [num_voxels, max_num_points_per_voxel, 3]
    gen_fea = torch.tensor(res['voxels'], dtype=torch.float32)
    # coors: [num_voxels, 3]
    gen_coords = torch.tensor(res['coordinates'], dtype=torch.int32)
    batch_idx = torch.zeros((gen_coords.shape[0], 1), dtype=torch.int32)  # batch=0
    gen_coords = torch.cat((batch_idx, gen_coords), dim=1)
    # # num_points: [num_voxels]
    gen_npoints = torch.tensor(res['num_points_per_voxel'], dtype=torch.int32)
    # 3. reconnect with compute graph
    gen_fea[:, :, :3] = ori_points[gen_fea[:, :, 3].long(), :]
    # # features[num_voxels, point_per_voxel, 3], N.B. set pos value to 0 in indice 0 (filling value)
    gen_fea = gen_fea[:, :, :3]
    # 4. you can do anything
    # voxel_vfe: [num_voxels, 3]
    voxel_vfe = torch.mean(gen_fea, dim=1)
    detached_vox = voxel_vfe.detach().numpy()
    spatial_shape = np.ceil(np.percentile(detached_vox, 100, axis=0) - np.percentile(detached_vox, 0, axis=0)).astype(np.int32)
    voxel_feature = spconv.SparseConvTensor(features=voxel_vfe, indices=gen_coords,
                                            spatial_shape=[50,50,50],
                                            batch_size=1)
```

**Comment by FrancescoMandru** [comment]
Hey @Cenbylin have you ever tried with a real point cloud and not a fake dataset? If you adapt your code with Semantic KITTI dataset you can throw it into the trash

**Comment by a-akram-98** [comment]
> generate batch of voxels requires cuda hashtable because dense table takes too much GPU memory, I plan to work on it after sparse convolution speed up, may be one month later.
> @Divadi Do you mean use VoxelGenerator as a differentiable layer? if so, I can finish it in this weekend.

@traveller59 Is the VoxelGenerator a differentiable layer !?

**Comment by zyl1336110861** [comment]
> Hi @Divadi I guess the use of NumPy (instead of Torch) is because of the convenience of c++ binding or operations on Cuda.
> 
> There is still a way to preserve gradient flow. It looks like to re-connect the new voxel tensor and the original compute graph.
> 
> You can try this:
> 
> ```python
> import spconv
> import torch
> # ------------------ #
> #  Data Preparation  #
> # ------------------ #
> ori_points = torch.randn(100, 3, requires_grad=True)
> voxel_generator = spconv.utils.VoxelGeneratorV2(voxel_size=[0.1] * 3,
>                                                 point_cloud_range=[-1, -1, -1, 1, 1, 1],
>                                                 max_num_points=50,
>                                                 max_voxels=8000,
>                                                 full_mean=False)
> # ------------------------------- #
> #  Associated with compute graph  #
> # ------------------------------- #
> # 1. retain raw indice. 
> ori_points_ = ori_points.detach()
> batch_point_indice = torch.arange(0, ori_points_.shape[0]).float().unsqueeze(1)  # [N, 1]
> # [N,3] -> [N,4], the add '1' is the indice in N
> ori_points_ = torch.cat((ori_points_, batch_point_indice), dim=1).view(-1, 4)
> 
> # 2. generate voxels
> # features[num_voxels, point_per_voxel, 3+1] 1(indice), coords: [num_voxels, 1+3] 1(batch)
> res= voxel_generator.generate(ori_points_.numpy())
> # features: [num_voxels, max_num_points_per_voxel, 3]
> gen_fea = torch.tensor(res['voxels'], dtype=torch.float32)
> # coors: [num_voxels, 1+3]
> gen_coords = torch.tensor(res['coordinates'], dtype=torch.int32)
> batch_idx = torch.zeros((gen_coords.shape[0], 1), dtype=torch.int32)  # batch=0
> gen_coords = torch.cat((batch_idx, gen_coords), dim=1)
> # num_points: [num_voxels]
> gen_npoints = torch.tensor(res['num_points_per_voxel'], dtype=torch.int32)
> 
> 
> # 3. reconnect with compute graph
> gen_fea[:,:,:3] = ori_points[gen_fea[:,:,3].long(),:]
> # features[num_voxels, point_per_voxel, 3], N.B. set pos value to 0 in indice 0 (filling value)
> gen_fea = gen_fea[:,:,:3]
> 
> 
> # 4. you can do anything
> # voxel_vfe: [num_voxels, 3]
> voxel_vfe = torch.mean(gen_fea, dim=1)
> voxel_feature = spconv.SparseConvTensor(features=voxel_vfe, indices=gen_coords, 
>                                         spatial_shape=[20, 20, 20], 
>                                         batch_size=1)
> simple_net = spconv.SparseConv3d(in_channels=3, out_channels=6,
>                                  kernel_size=3, stride=2)
> output = simple_net(voxel_feature).dense()
> 
> # ------ #
> #  Test  #
> # ------ #
> loss = output.sum()
> print('previous:', ori_points.grad)
> loss.backward()
> print('current:', ori_points.grad)
> ```
> 
> Hope it can help you.
> 
> There is another concern: the points in voxels didn't necessarily cover all original points. So the supervising signal is also sparse (i.e. many zeros in ori_points.grad).
> 
> Are you training a point cloud GAN? If so, could we discuss them a little bit?

Thank you for your solution! But when I try to do this with spconv2.1 with Point2Voxel function, I get the two grad results with None. I don't know the difference between the VoxelGenerator and Point2Voxel.
